File: CarFactory.java
public class CarFactory {

    public Car createCar(Car carType){
        if (carType==null){
            return null;
        }
        //TODO: Ask about should a simple factory be used or factory method or Abstract factory

        if(carType instanceof ElectricCar){
            return new ElectricCar(carType.getVin(),carType.getBrand(),carType.getModel(), carType.getYearMade(), carType.getPrice());
        } else if (carType instanceof PetrolCar) {
            //TODO: adjust the typecast
            // I apologise for the typecast but it *should work*
            return new PetrolCar(carType.getVin(),carType.getBrand(),carType.getModel(), carType.getYearMade(), carType.getPrice());
        } else if (carType instanceof HydrogenCar){
            return new HydrogenCar(carType.getVin(),carType.getBrand(),carType.getModel(), carType.getYearMade(), carType.getPrice());
        } else return null;

    }
}

File: CarFactory.cs
using System;

public class CarFactory
{
    public Car CreateCar(Car carType)
    {
        if (carType == null)
        {
            return null;
        }

        // TODO: Ask about whether a simple factory, factory method, or abstract factory should be used.

        if (carType is ElectricCar)
        {
            return new ElectricCar(carType.Vin, carType.Brand, carType.Model, carType.YearMade, carType.Price);
        }
        else if (carType is PetrolCar)
        {
            return new PetrolCar(carType.Vin, carType.Brand, carType.Model, carType.YearMade, carType.Price);
        }
        else if (carType is HydrogenCar)
        {
            return new HydrogenCar(carType.Vin, carType.Brand, carType.Model, carType.YearMade, carType.Price);
        }
        else
        {
            return null;
        }
    }
}
